src/main/resources/resource/WebGui/app/service/js/WebkitSpeechRecognitionGui.js

```diff
@@ -1,239 +1,290 @@
+        "Українська": "uk-UA",
+        "български": "bg-BG",
+        "ภาษาไทย": "th-TH",
+        "中文 - 中文 (台灣)": "cmn-Hant-TW",
+        "中文 - 普通话 (中国大陆)": "cmn-Hans-CN",
+        "中文 - 普通话 (香港)": "cmn-Hans-HK",
+        "中文 - 粵語 (香港)": "yue-Hant-HK",
+        "日本語": "ja-JP",
+        "한국어": "ko-KR",
     }
-    ;
-    // when to use $scope or anything?!
-    $scope.currResponse = '';
-    $scope.utterance = '';
-    $scope.current_text = '';
-    // start info status
-    $scope.status = 'Click on the microphone icon and begin speaking.[init]';
-    $scope.rows = [];
-    // current state for the button i guess?
-    $scope.recognizing = false;
-    // always grab the right service!
-    $scope.service = mrl.getService($scope.service.name);
-    // webkit $scope.recognition google speech
-    // Check if webkitSpeechRecognition exists
-    if (!('webkitSpeechRecognition'in window)) {
-        $scope.wkavailable = false;
-        $log.info('WebkitSpeechRecognition', 'not available');
-        $scope.status = 'WebkitSpeechRecognition not available. Need to use Chrome on a supported platform.';
-    } else {
-        $scope.wkavailable = true;
-        $scope.recognition = new webkitSpeechRecognition();
-        // config properties on the webkit speech stuff.
-        $scope.recognition.continuous = true;
-        $scope.recognition.interimResults = true;
-        mrl.sendTo($scope.service.name, "startListening");
-        // called when $scope.recognition starts.
-        $scope.recognition.onstart = function() {
-            $scope.started = true;
-            $scope.recognizing = true;
-            $scope.status = 'Speak now.';
-            $scope.listenbuttonimg = 'service/img/mic-animate.gif';
-            $scope.$apply();
-            mrl.sendTo($scope.service.name, "listeningEvent", "true");
+
+    let recognizer = null
+
+    $scope.restartCnt = 0
+    $scope.interimTranscript = ''
+    $scope.publishedText = ''
+
+    // corresponds to internal RecognizedResult class
+    // in AbstractSpeechREcognizer
+    $scope.recognizedResult = {
+        text: null,
+        confidence: null,
+        isFinal: false
+    }
+
+    $scope.selectedLanguage = "en-US"
+    $scope.startTimestamp = null
+    $scope.stopRequested = false
+    $scope.errorText = null
+    $scope.log = []
+
+    // this is really a js service
+    // and this is the initial service state we want
+    $scope.service = {
+        isListening: false,
+        status: null,
+        img: '../WebkitSpeechRecognition/mic.png',
+        webkitSupport: true
+    }
+
+    $scope.changeListeningState = function() {
+        if (!$scope.service.isListening) {
+            $scope.setState('start')
+        } else {
+            $scope.setState('stop')
         }
     }
-    ;if ($scope.wkavailable) {
-        // called when there's an error (handles a few cases)
-        $scope.recognition.onerror = function(event) {
-            if (event.error == 'no-speech') {
```

Comment:
please restore this if block to make sure the proper error messages and actions are exposed to end users.

Reply:
Sure, I'd like to report the proper error states to the user, but its more complex than just slapping this code back in.  There were lots of problems, and errors states it was not handling either - for example the test for appropriate webkit support wasn't being done, additionally - the user was not being notified if webkit or speechrecognition was not supported.   We need the ability to show "recording" vs "listening" - it becomes important for privacy and the ability to support a wake word.
Working on these parts currently

Reply:
Back on October this was working great.  At a minimum, it needs to detect 
1. when no microphone is found.
2. when the microphone is denied access.   In either of those scenarios it must not attempt to re-listen again.. because those are non-recoverable errors.  
current behavior ultimately will crash and lockup chromium on raspi how it's implemented, and never let the end user know what the underlying cause was.

Reply:
yup - agree

Reply:
All error information is propagated to UI, nothing is hidden.  Only "high rate" is appended when appropriate.   All errors stop and do not try to restart, with the exception of "no-speech" which is not an error that should be stopped on.  This change should prevent the tight endless loop which you noticed previously.

Reply:
Tested, verified.